test/integration/integration_test.go

```diff
@@ -7,54 +7,1039 @@
+			review := makeTokenReviewRequest(apiserverToken, token, testEnv.OIDCServerCA(), false)
+
+			Expect(review.Status.Authenticated).To(BeFalse())
+			Expect(review.Status.User.Username).To(BeEmpty())
+			Expect(review.Status.User.Groups).To(BeEmpty())
+
+			waitForOIDCResourceToBeDeleted(ctx, k8sClient, provider)
+		})
+
+		It("Should authenticate user against the target audience", func() {
+			idp := createAndStartIDPServer(1)
+			defer stopIDP(ctx, idp)
+
+			userPrefix1 := "userpref1:"
+			userPrefix2 := "userpref2:"
+			userClaim := "custom-sub"
+			groupsPrefix1 := "grouppref1:"
+			groupsPrefix2 := "grouppref2:"
+			groupsClaim := "custom-groups"
+			provider1 := defaultOIDCProvider(fmt.Sprintf("https://localhost:%v", idp.ServerSecurePort), idp.CA())
+			provider1.Spec.UsernameClaim = &userClaim
+			provider1.Spec.UsernamePrefix = &userPrefix1
+			provider1.Spec.GroupsClaim = &groupsClaim
+			provider1.Spec.GroupsPrefix = &groupsPrefix1
+			provider1.Spec.ClientID = "client1"
+
+			waitForOIDCResourceToBeCreated(ctx, k8sClient, provider1)
+
+			provider2 := defaultOIDCProvider(fmt.Sprintf("https://localhost:%v", idp.ServerSecurePort), idp.CA())
+			provider2.Spec.UsernameClaim = &userClaim
+			provider2.Spec.UsernamePrefix = &userPrefix2
+			provider2.Spec.GroupsClaim = &groupsClaim
+			provider2.Spec.GroupsPrefix = &groupsPrefix2
+			provider2.Spec.ClientID = "client2"
+
+			waitForOIDCResourceToBeCreated(ctx, k8sClient, provider2)
+
+			user := "this-is-my-identity"
+			claims1 := defaultClaims()
+			claims1[userClaim] = user
+			claims1[groupsClaim] = []string{"admin"}
+			claims1["iss"] = fmt.Sprintf("https://localhost:%v", idp.ServerSecurePort)
+			claims1["aud"] = "client1"
+
+			token1, err := idp.Sign(0, claims1)
+			Expect(err).NotTo(HaveOccurred())
+
+			claims2 := defaultClaims()
+			claims2[userClaim] = user
+			claims2[groupsClaim] = []string{"admin"}
+			claims2["iss"] = fmt.Sprintf("https://localhost:%v", idp.ServerSecurePort)
+			claims2["aud"] = "client2"
+
+			token2, err := idp.Sign(0, claims2)
+			Expect(err).NotTo(HaveOccurred())
+
+			review1 := makeTokenReviewRequest(apiserverToken, token1, testEnv.OIDCServerCA(), true)
+
+			Expect(review1.Status.Authenticated).To(BeTrue())
+			Expect(review1.Status.User.Username).To(Equal(userPrefix1 + user))
+			Expect(review1.Status.User.Groups).To(ConsistOf(groupsPrefix1 + "admin"))
+
+			review2 := makeTokenReviewRequest(apiserverToken, token2, testEnv.OIDCServerCA(), true)
+
+			Expect(review2.Status.Authenticated).To(BeTrue())
+			Expect(review2.Status.User.Username).To(Equal(userPrefix2 + user))
+			Expect(review2.Status.User.Groups).To(ConsistOf(groupsPrefix2 + "admin"))
+
+			waitForOIDCResourceToBeDeleted(ctx, k8sClient, provider1)
+			waitForOIDCResourceToBeDeleted(ctx, k8sClient, provider2)
+		})
 	})
 })
+
+func waitForOIDCResourceToBeCreated(ctx context.Context, k8sClient client.Client, oidc *authenticationv1alpha1.OpenIDConnect) {
+	err := k8sClient.Create(ctx, oidc)
```

Comment:
Simiilar, if the resource already exist, the Create call will fail. One option is to use  `sigs.k8s.io/controller-runtime/pkg/controller/controllerutil.CreateOrUpdate`.

Or maybe it is desired the test to fail?

Reply:
Default resources are generated with random names and we do not expect that a resource will be already present. If a resource already exists when calling this function something undesired has happened and it is better for the test to fail.